install/bootupd: chroot to deployment

jbtrystram opened the pull request:
When `--src-imgref` is passed, the deployed systemd does not match the running environnement. In this case, let's chroot into the deployment before calling bootupd. This makes sure we are using the binaries shipped in the image (and relevant config files such as grub fragements).

We could do that in all cases but i kept it behind the `--src-imgref` option since when using the target container as the buildroot it will have no impact, and we expect this scenario to be the most common.

In CoreOS we have a specific test that checks if the bootloader was installed with the `grub2-install` of the image.

Fixes https://github.com/bootc-dev/bootc/issues/1559 Also see https://github.com/bootc-dev/bootc/issues/1455

jbtrystram commented:
/gemini review

cgwalters commented:
OK there's some legit failures here like `  content: error: boot data installation failed: installing component EFI: Listing partitions of /dev/loop0: No such file or directory (os error 2)`.

cgwalters reviewed (changes requested):
Marking as requested changes due to failing CI

cgwalters commented on crates/lib/src/bootloader.rs:
```diff
@@ -50,22 +52,70 @@ pub(crate) fn install_via_bootupd(
+    // let's bind mount it to a temp mountpoint under /run
+    // so it gets carried over in the chroot.
+
+    let rootfs_mount = if rootfs.starts_with("/run") {
```
Thinking about this it may be a lot simpler to just setup the API filesystems by default in the install flow.

(and per discussion mount e.g. `/boot` in the target rootfs)

jbtrystram commented:
> Note that it's only Fedora variants that are failing; should reproduce locally via e.g. `just base=quay.io/fedora/fedora-bootc:42 test-tmt install-unified`

Yes, i can reproduce that locally ! Ok so I dug deeper and I think I figured out something :
These tmt test run the install process through a [systemd transient unit with `MountFlags=Slave`](https://github.com/bootc-dev/bootc/blob/64ad5c539f1fd779fe7b4887cbf681108fc352d8/tmt/tests/booted/test-install-unified-flag.nu#L30), which cause the `/dev/loop` device to not get mounted inside the chroot target. 

I am not sure what happens because removing the `MountFlags` does not fix it, but removing the systemd-run wrapper does help : 

```
truncate -s 10G disk.img
systemd-run  -qdPG -- /bin/sh -c $"./bootc install to-disk --disable-selinux --via-loopback --filesystem xfs  --source-imgref docker://quay.io/centos-bootc/centos-bootc:stream10 ./disk.img"
....
Installing bootloader via bootupd
error: boot data installation failed: installing component EFI: Listing partitions of /dev/loop0: No such file or directory (os error 2)
```

Without the systemd wrapper : 
```
truncate -s 10G disk.img
./bootc install to-disk --disable-selinux --via-loopback --filesystem xfs  --source-imgref docker://quay.io/centos-bootc/centos-bootc:stream10 ./disk.img
.....
Bootloader: grub
Installing bootloader via bootupd
Added 01_users.cfg
Added 10_blscfg.cfg
Added 14_menu_show_once.cfg
Added 30_uefi-firmware.cfg
Added 41_custom.cfg
Installed: grub.cfg
Installed: bootuuid.cfg
Installed: "centos/grub.cfg"
Installed: "centos/bootuuid.cfg"
```

In the second case the chroot works.

Another thing to note, and i cannot figure this out yet : the error finding the `/dev/loop0` device is yielded from [bootc code](https://github.com/bootc-dev/bootc/blob/64ad5c539f1fd779fe7b4887cbf681108fc352d8/crates/blockdev/src/blockdev.rs#L221-L227) and not bootupd

jbtrystram commented:
Thinking more about this : in both cases the mounting and formating of the block device (`disk.img`) through the loopback device works. So it's systemd not letting us bind mount inside the chroot ?

jbtrystram commented:
Ok so after doing a bunch of testing around `systemd-run` it's not the cause. 
I ended up changing the `bootloader.rs` code to simply do : `chroot <deployment_path> /usr/sbin/sfdisk -J /dev/loop0` and it works !? But it does not work without specifying `sfdisk` full path. So maybe it's down to a `$PATH` issue ?

I will try to add the full path but that requires a rebuild of bootupd as well, so that's for tomorrow. 

edit: ok I did ask bootc to run `which sfdisk` in the chroot and sure enough: `no sfdisk in (/usr/local/bin:/usr/bin)`

jbtrystram commented on crates/lib/src/bootloader.rs:
```diff
@@ -50,22 +52,70 @@ pub(crate) fn install_via_bootupd(
+        }
+        // Append the `bootupctl` command, it will be passed as
+        // an argument to chroot
+        vec![target_root.as_str(), "bootupctl"]
```
we would then need to have another bind mount `rootfs/boot` into `deployment_path/boot` I think.

jbtrystram commented:
/gemini review

jbtrystram commented on crates/lib/src/bootloader.rs:
```diff
@@ -91,22 +91,67 @@ pub(crate) fn install_via_bootupd(
+        for partition in &device.partitions {
+            cmd = cmd.bind_device(&partition.node);
+        }
+        // // TODO : is it needed ?
```
So this is a leftover from the manual `chroot`, but I noticed that bwrap does set this up for us so I commented it out.

You're right about other things that could leak in though

jbtrystram commented:
> There was one failure in fedora-43/ostree which _might_ have been related to this; I restarted it. Clearly one thing that would help again is for us to lock the blockdev and pass it down as a fd and not a path through the stack.

I think that's because of the commented out $PATH. I'm testing locally to verify.

jbtrystram commented:
Ok, so the fix is to always inject a reasonnable default path. I really don't understand why though.
I started multiples images and VMs to try to figure it out.
In the fedora case, `sfdisk` path cannot be resolved, although it's installed in `/usr/bin` and all symlinks point to there anyway. I suppose that's because of the [sbin and bin](https://fedoraproject.org/wiki/Changes/Unify_bin_and_sbin) merge but since this is all symlinks pointing to the same place, I don't see how the value of $PATH have an impact.
